Rely on iconv and symfony/polyfill-*

nicolas-grekas opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #16240 |
| License | MIT |
| Doc PR | - |

Status: needs work
symfony/polyfill-\* packages are not ready yet. Still, review welcomed.

nicolas-grekas committed:
Rely on iconv and symfony/polyfill-*

xabbuh commented:
The DoctrineBridge needs an update of the `composer.json` file.

nicolas-grekas commented:
@xabbuh not sure: iconv is an implicit requirement now (same as ext-json)

xabbuh commented on src/Symfony/Component/Security/Core/composer.json:
```diff
@@ -17,25 +17,25 @@
     ],
     "require": {
         "php": ">=5.3.9",
-        "paragonie/random_compat": "~1.0"
```
This must be kept (`random_bytes()` is used in the `SecureRandom` class).

nicolas-grekas replied:
yep, I'm going to add random_compat to polyfill-php70 that's why it's removed from here

xabbuh replied:
okay

xabbuh commented:
@nicolas-grekas Is iconv even available on Windows by default?

nicolas-grekas commented:
@xabbuh yes, it's bundled on Windows without adding any dll, at least in the binaries provided on windows.php.net

Tobion commented:
Is the idea also to move the Intl component to Polyfill? Isn't the Intl component not basically a Polyfill or does it do more?

Tobion commented on src/Symfony/Component/Yaml/Inline.php:
```diff
@@ -52,7 +52,7 @@ public static function parse($value, $exceptionOnInvalidType = false, $objectSup
         }
 
-        if (function_exists('mb_internal_encoding') && ((int) ini_get('mbstring.func_overload')) & 2) {
+        if (2 /* MB_OVERLOAD_STRING */ & (int) ini_get('mbstring.func_overload')) {
```
isn't this constant polyfilled?

stof replied:
shouldn't we check for ``extension_loaded('mbstring')`` here ? Only the extension itself can overload functions anyway

nicolas-grekas replied:
the extension must be enabled for ini_get to return `2` so this is equivalent

nicolas-grekas replied:
See https://3v4l.org/aBCev

nicolas-grekas commented:
@Tobion there where a few missing require for symfony/polyfill-intl-icu that I fixed. I looked at symfony/intl and it's both a shim and a utility component. I added `@internal` tags to the shim-related classes. Moving the code to polyfill-intl-icu could be useful as a cleanup but it's a bigger work and it won't provide much practical value so I'd like to postpone it for later if we want it.

nicolas-grekas commented:
Tests are green! (appveyor failure is a transient test)
Status: needs review

Tobion commented on src/Symfony/Component/Templating/PhpEngine.php:
```diff
@@ -330,6 +330,9 @@ public function escape($value, $context = 'html')
      */
     public function setCharset($charset)
     {
+        if ('UTF8' === $charset = strtoupper($charset)) {
```
why do we need that? Shouldn't it be deprecated if it is for bc? Looks like a hack.

nicolas-grekas replied:
it's because iconv on windows doesn't accept utf8 as an encoding name: only utf-8 is supported

Tobion replied:
would be good to add this as comment

nicolas-grekas replied:
comment added

Tobion commented on composer.json:
```diff
@@ -96,10 +101,8 @@
-            "src/Symfony/Component/HttpFoundation/Resources/stubs",
             "src/Symfony/Component/Intl/Resources/stubs"
-        ],
-        "files": [ "src/Symfony/Component/Intl/Resources/stubs/functions.php" ]
```
this file should be removed actually

nicolas-grekas replied:
good catch, removed

Tobion commented:
:+1: 

Status: Reviewed

fabpot commented:
Thank you @nicolas-grekas.

JHGitty commented:
@fabpot Merging it was not a good idea yet. #16363

Tobion commented:
I merged 2.8 in master and resolved all the conflicts.

Tobion commented on src/Symfony/Component/Serializer/composer.json:
```diff
@@ -16,7 +16,8 @@
     "require": {
-        "php": ">=5.3.9"
+        "php": ">=5.3.9",
+        "symfony/polyfill-php54": "~1.0"
```
This is wrong: #16382
